### ISSUE TITLE
Deadlock in DefaultStateMachineService

### ISSUE THREAD

**Issue by hmeerlo** [spring-projects/spring-statemachine#770]
I believe there is a deadlock situation in the DefaultStateMachineService in the acquireStatemachine method:

```java
        @Override
	public StateMachine<S, E> acquireStateMachine(String machineId, boolean start) {
		log.info("Acquiring machine with id " + machineId);
		synchronized (machines) {
			StateMachine<S,E> stateMachine = machines.get(machineId);
			if (stateMachine == null) {
				log.info("Getting new machine from factory with id " + machineId);
				stateMachine = stateMachineFactory.getStateMachine(machineId);
				if (stateMachinePersist != null) {
					try {
						StateMachineContext<S, E> stateMachineContext = stateMachinePersist.read(machineId);
						stateMachine = restoreStateMachine(stateMachine, stateMachineContext);
					} catch (Exception e) {
						log.error("Error handling context", e);
						throw new StateMachineException("Unable to read context from store", e);
					}
				}
				machines.put(machineId, stateMachine);
			}
			return handleStart(stateMachine, start);
		}
	}
```

In my opinion the `handleStart` call at the end should be outside of the `synchronized(machines)` block. When I now start my state machines and one of them hangs in a very long transition, I can not acquire a new statemachine at all. I see no reason why the `handleStart` should be inside the `synchronized` call

**Comment by jvalkeal** [comment]
Yes, you do have a point here. Implementation is indeed way too naive.

**Comment by jvalkeal** [comment]
@hmeerlo lemmy know if fix 0a760f5bc049922a262d8a34f0027f886f888ec1 in `2.1.x` works for you.

**Comment by hmeerlo** [comment]
@jvalkeal 👍 that is exactly what I have done locally and that works fine
